### PR TITLE
feat(contributing): add 'topic-namespaces.md' in ros-nodes

### PR DESCRIPTION
**Pull request by cyn-liu**
Signed-off-by: Cynthia Liu <cynthia.liu@autocore.ai>

## Description
This pr add topic-namespaces.md in autoware documentation.
related issue https://github.com/autowarefoundation/autoware-documentation/issues/222.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The Reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by cyn-liu** [comment]
@kenji-miyake  Hello! Do you have any suggestions on how to modify this PR?

I have some doubts about the guides in this directory of `/docs/contributing/coding-guildlines/ros-nodes`, which can be found in the official ROS tutorials. I see that you have not agreed on the relevant PR ( [pr220](https://github.com/autowarefoundation/autoware-documentation/pull/220),[pr216](https://github.com/autowarefoundation/autoware-documentation/pull/216)), so do you have any suggestions on how to write each guide in this directory?

**Comment by kenji-miyake** [comment]
@cyn-liu Thank you for the confirmation.

In this directory, we'd like to put documents about `How to design Autoware's ROS nodes`.
And general ROS 2 design guidelines will be linked from the documents if necessary.

I'm planning to propose each document, but it takes more time because my resource is currently very limited.
I'm really sorry about that. Could you wait for a while, please?
Thank you for your patience. :bow:

**Comment by cyn-liu** [comment]
@kenji-miyake Thanks for your answer, I can wait for a while.

I plan to write the documentation for the  `ros-node` section , so I would like to get more suggestions to help me write it well.

**Comment by kenji-miyake** [comment]
@cyn-liu I recommend you to read the design section and understand the current Autoware and future plans well! The contribution guidelines are just a summary of them.
https://autowarefoundation.github.io/autoware-documentation/main/design/

File: docs/contributing/coding-guidelines/ros-nodes/topic-namespaces.md [hunk 1]
```diff
@@ -1,5 +1,70 @@
...
+    <remap from="output/shift" to="$(var output_shift)"/>
+    <remap from="output/turn_signal" to="$(var output_turn_signal)"/>
+    <remap from="output/gate_mode" to="$(var output_gate_mode)"/>
+    <remap from="output/heartbeat" to="$(var output_heartbeat)"/>
+    <remap from="output/vehicle_engage" to="$(var output_vehicle_engage)"/>
+  </node>
+</launch>
+```
+
+## Topic names in the code
```
Comment (mitsudome-r) [inline docs/contributing/coding-guidelines/ros-nodes/topic-namespaces.md:63]: Maybe it's better to add example code snippet as well (e.g., pub/sub)

File: docs/contributing/coding-guidelines/ros-nodes/topic-namespaces.md [hunk 2]
```diff
@@ -1,5 +1,99 @@
...
+- sensing
+- vehicle
+- map
+- system
+
+When a node is run in a namespace, all topics which that node publishes are given that same namespace. All nodes in the Autoware stack must support namespaces by avoiding practices such as publishing topics in the global namespace.
+
+In general, topics should be namespaced based on the function of the node which produces them and not the node (or nodes) which consume them.
+
+Classify topics as input or output topics based on they are subscribed or published by the node. In the node, input topic is named `input/topic_name` and output topic is named `output/topic_name`.
```
Comment (xmfcx) [inline docs/contributing/coding-guidelines/ros-nodes/topic-namespaces.md:31]:
```suggestion
Classify topics as input or output topics based on they are subscribed or published by the node. In the node, input topic is named `~/sub/topic_name` and output topic is named `~/pub/topic_name`.
```
As discussed in https://github.com/orgs/autowarefoundation/discussions/3171

File: docs/contributing/coding-guidelines/ros-nodes/topic-namespaces.md [hunk 3]
```diff
@@ -1,5 +1,99 @@
...
+
+    <remap from="output/control_command" to="$(var output_control_command)"/>
+    <remap from="output/external_control_command" to="$(var output_external_control_command)"/>
+    <remap from="output/shift" to="$(var output_shift)"/>
+    <remap from="output/turn_signal" to="$(var output_turn_signal)"/>
+    <remap from="output/gate_mode" to="$(var output_gate_mode)"/>
+    <remap from="output/heartbeat" to="$(var output_heartbeat)"/>
+    <remap from="output/vehicle_engage" to="$(var output_vehicle_engage)"/>
+  </node>
+</launch>
```
Comment (xmfcx) [inline docs/contributing/coding-guidelines/ros-nodes/topic-namespaces.md:60]:
```suggestion
<launch>
  <arg name="sub/foo" default="/path/to/foo"/>
  <arg name="pub/bar" default="/path/to/bar"/>
  <arg name="cli/baz" default="/path/to/baz"/>
  <arg name="srv/qux" default="/path/to/qux"/>

  <node pkg="package_name" exec="package_executable" name="node_name" output="screen">
    <remap from="~/sub/foo" to="$(var sub/foo)"/>
    <remap from="~/pub/bar" to="$(var pub/bar)"/>
    <remap from="~/cli/baz" to="$(var cli/baz)"/>
    <remap from="~/srv/qux" to="$(var srv/qux)"/>
  </node>
</launch>
```

Also taking https://github.com/orgs/autowarefoundation/discussions/3171#discussioncomment-4597052 as recommendation.

File: docs/contributing/coding-guidelines/ros-nodes/topic-namespaces.md [hunk 4]
```diff
@@ -1,5 +1,99 @@
...
+- map
+- system
+
+When a node is run in a namespace, all topics which that node publishes are given that same namespace. All nodes in the Autoware stack must support namespaces by avoiding practices such as publishing topics in the global namespace.
+
+In general, topics should be namespaced based on the function of the node which produces them and not the node (or nodes) which consume them.
+
+Classify topics as input or output topics based on they are subscribed or published by the node. In the node, input topic is named `input/topic_name` and output topic is named `output/topic_name`.
+
+Configure the topic in the node's launch file. Take the `joy_controller` node as an example, in the following example, set the input and output topics and remap topics in the `joy_controller.luanch.xml` file.
```
Comment (xmfcx) [inline docs/contributing/coding-guidelines/ros-nodes/topic-namespaces.md:33]:
```suggestion
Here is an example launch file. The publisher, subscriber, service and the client are remapped.
```

**Comment by xmfcx** [comment]
I will merge this as it is and propose an updated PR including the discussion in https://github.com/orgs/autowarefoundation/discussions/3171

I think this will be simpler because we are planning to change the existing naming scheme.

@cyn-liu this PR covers the existing naming scheme well. Thanks for creating this document!

**Review by xmfcx — APPROVED** [review]

**Comment by xmfcx** [comment]
Created the issue in https://github.com/autowarefoundation/autoware-documentation/issues/292 until we have a bit more progress in the discussion.

**Pull request merged**